### DISCUSSION TITLE
How to cancel a long-running request?

### DISCUSSION THREAD

**Discussion by harvister** [oakserver/oak#322] [Q&A]
if I set `context.request.body` equal to a `Deno.Reader` in order to stream a long-running response, how do I just cancel it half way through?

I'm looking for something like `context.cancel()`

I'm running into a problem where a `fetch` call from a browser client will not fail until either the full request body is read, or the server is stopped.  I want to be able to tell the client to fail the request without having to read the remaining request body, which I'm also reading as a `Deno.Reader`.  If the request body is a 1 GB file, I do not want to have to read all of it just so the client will fail the fetch request.

**Comment by kitsonk** [comment]
Once you have responded, with a `Deno.Reader` it is effectively outside of the hands of oak at that point and it is handed back to the server at that point.

You need to finish the body in some fashion, so you need a `Deno.Reader` that will return a `null` indicating an EOF whenever whatever condition is met.  I haven't tried it but I suspect if you `.close()` a `Deno.File` it should cause the next read to be a `null`, but otherwise you would need to implement something that provides a `Deno.Reader` interface but supports cancelling/closing and then hold on to a reference to that when you send it back as a body.

**Reply by harvister** [reply]
I agree, and I will be going this route.

However, there is an edge case I encountered which kind of led me to asking about this...

After setting the `context.response.body` to my `Deno.Reader`, and sending some initial data, if the consumer of that reader, in this case the browser fetch client, for whatever reason stops accepting any more data (like when the browser fetch client first expects its entire request body to be read before it will read the response body), the reader's `read` method will not be called again, and thus cannot be errored, or given a `null` to indicate EOF, leaving the request open until the server is stopped.

Yes, the workaround in this particular case is to just ensure you read the full request body, and all is well.  However, I am adding a feature to allow very large file uploads, and thus would like to avoid reading large request bodies only to say it errored.  I can also add timeouts to the client, but in some cases I won't be controlling every client accessing this.
